File: 25.reverse-nodes-in-k-group.js
/*
 * @lc app=leetcode id=25 lang=javascript
 *
 * [25] Reverse Nodes in k-Group
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
  let stack = []
  let currHead = prevHead = head
  while (currHead) {
    stack.push(currHead.val)
    if (stack.length >= k) {
      while (stack.length) {
        prevHead.val = stack.pop()
        prevHead = prevHead.next
      }
    }
    currHead = currHead.next
  }
  return head
  // iterate k times with currHead, store in stack
  // when stack.length === k, pop stack
  // until empty storing values in prevHead
};
// @lc code=end

File: cpp/25.reverse-nodes-in-k-group.cpp
/*
 * @lc app=leetcode id=25 lang=cpp
 *
 * [25] Reverse Nodes in k-Group
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        // iterate k elements in linked list
        // store elements in stack
        // pop stack and replace values
        // O(n) time O(k) space
        ListNode *headPtr = head;
        while (head)
        {
            stack<int> stacker;
            ListNode *currNode = head;
            for (int i = 0; i < k; i++)
            {
                if (!currNode)
                {
                    return headPtr;
                }
                stacker.push(currNode->val);
                currNode = currNode->next;
            }
            while (!stacker.empty())
            {
                head->val = stacker.top();
                stacker.pop();
                head = head->next;
            }
        }
        return headPtr;
    }
};
// @lc code=end
